### DISCUSSION TITLE
issue: Duplicate fields after deleting. useFieldArray and react-query

### DISCUSSION THREAD

**Discussion by xiaohai-huang** [react-hook-form/react-hook-form#12233] [Q&A]
### Version Number

7.52.1

### Codesandbox/Expo snack

https://stackblitz.com/edit/vitejs-vite-sjvkpq?file=src%2FApp.jsx

### Steps to reproduce

1. Go to https://stackblitz.com/edit/vitejs-vite-sjvkpq?file=src%2FApp.jsx
2. Click on the "New" button, enter some text at the input field then click the "Save" button.
3. Repeat the above step to create another todo item.
4. Click the "Delete" button at the first todo item.
5. The first todo item's text is changed to the second todo's text

[useFieldArray-bug.webm](https://github.com/react-hook-form/react-hook-form/assets/56380096/3a747ed7-01ed-41e0-8194-79c3c1f38097)

### Expected behaviour

The first todo item should be deleted instead of using the text from the next todo item.

### What browsers are you seeing the problem on?

Chrome, Edge

### Relevant log output

```shell
Sorry for not using the __codesandbox__ template. The react version of that template is low that doesn't support `react-query`, that's why I used __stackblitz__.
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by abnud11** [comment]
@xiaohai-huang The problem doesn't seem to be about `useFieldArray`, if you inspect what's returned from `watch` which is returned by `useForm` you'll see that it prints wrong state too.

It seems to be an issue of `useForm` not syncing its state with `values` passed to it, it should update its state when `values` update.

Interesting finding: if you remove the `keepDirtyValues: true` option your example actually works.

In my opinion we either have a bug or a documentation problem, I really don't understand this `keepDirtyValues`, or maybe it's the combination of `values` and `keepDirtyValues` is what I don't understand.

**Comment by abnud11** [comment]
On second thought, it's the usage of an array field with dirty fields state I don't understand, when inspecting `dirtyFields` of `formState` I get two array items when adding and saving two items, it looks like this:

```json
{
  "todos": [
    {
      "uid": false,
      "text": false,
      "checked": false
    },
    {
      "uid": true,
      "text": true,
      "checked": true
    }
  ]
}
```
While `values` is like this:
```json
{
  "values": {
    "todos": [
      {
        "uid": "0.36061416878791586",
        "text": "234234",
        "checked": false
      },
      {
        "uid": "0.32548578068248113",
        "text": "2342626",
        "checked": false
      }
    ]
  }
}
```
See my updated example which shows multiple inspections of `values`, `watch`, `dirtyFields` and `fields` from `useFieldsArray`
https://stackblitz.com/edit/vitejs-vite-qu6gpp?file=src%2FApp.jsx

**Comment by xiaohai-huang** [comment]
@abnud11 Thank you.
> when inspecting `dirtyFields` of `formState` I get two array items when adding and saving two items, it looks like this

I expect the `dirtyFields` to be all `false` like below after saving two items. I am wondering if it is okay to use `useFieldArray` with `resetOptions: { keepDirtyValues: true }`? By the way, I use react-query to fetch data and update the form if someone else updated the form, that's why I need `keepDirtyValues`.

```js
{
  "todos": [
    {
      "uid": false,
      "text": false,
      "checked": false
    },
    {
      "uid": false,
      "text": false,
      "checked": false
    }
  ]
}
```

**Comment by abnud11** [comment]
@xiaohai-huang Why is false expected for all fields? I understand `dirtyFields` value when form state is an object but don't understand how it works for array fields.

**Comment by xiaohai-huang** [comment]
@abnud11 Well, I am not an expert in `dirtyFields` as well. My understanding is that, after clicking the __Save__ button all the todos are saved at database, none of the fields should be dirty.

**Comment by Balour** [comment]
I was seeing the same behavior as you but using 7.45.1 in certain conditions. If the record was not from the DB I would just call remove(index) and that works, but asking the user and reacting on the answer via useEffect it would do this behavior. It's like if the "remove" being called  in the useEffect didn't react the same as the one directly in the function. I removed this old dialog that needed a useEffect in favor of a proper dialog component with "onConfirm" props and now remove(index) works like a charm

Hope it can help or inspire your debugging
